### PR TITLE
[SPARK-9544] [MLLIB] add Python API for RFormula

### PR DESCRIPTION
**Pull request by mengxr**
Add Python API for RFormula. Similar to other feature transformers in Python. This is just a thin wrapper over the Scala implementation. @ericl @MechCoder

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala [hunk 1]
```diff
@@ -63,31 +61,26 @@ class RFormula(override val uid: String) extends Estimator[RFormulaModel] with R
    */
   val formula: Param[String] = new Param(this, "formula", "R model formula")
 
-  private var parsedFormula: Option[ParsedRFormula] = None
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: I removed it because there are multiple places to set params. We need to parse the input formula just-in-time.

File: python/pyspark/ml/feature.py [hunk 2]
```diff
@@ -1110,6 +1110,80 @@ class PCAModel(JavaModel):
...
+        setParams(self, formula=None, featuresCol="features", labelCol="label")
+        Sets params for RFormula.
+        """
+        kwargs = self.setParams._input_kwargs
+        return self._set(**kwargs)
+
+    def setFormula(self, value):
+        """
+        Sets the value of :py:attr:`formula`.
+        """
```
Comment (MechCoder) [inline python/pyspark/ml/feature.py]: Is it useful to check if value is a string or not?

File: python/pyspark/ml/feature.py [hunk 3]
```diff
@@ -1110,6 +1110,80 @@ class PCAModel(JavaModel):
...
 
 
+@inherit_doc
+class RFormula(JavaEstimator, HasFeaturesCol, HasLabelCol):
+    """
+    .. note:: Experimental
+
+    Implements the transforms required for fitting a dataset against an
+    R model formula. Currently we support a limited subset of the R
+    operators, including '~', '+', '-', and '.'. Also see the R formula
```
Comment (MechCoder) [inline python/pyspark/ml/feature.py]: Would it be nice to have an example about the dot operator? I could not find one in the RFormulaSuite as well.

File: python/pyspark/ml/feature.py [hunk 4]
```diff
@@ -1110,6 +1110,80 @@ class PCAModel(JavaModel):
...
+    docs:
+    http://stat.ethz.ch/R-manual/R-patched/library/stats/html/formula.html
+
+    >>> df = sqlContext.createDataFrame([
+    ...     (1.0, 1.0, "a"),
+    ...     (0.0, 2.0, "b"),
+    ...     (0.0, 0.0, "a")
+    ... ], ["y", "x", "s"])
+    >>> rf = RFormula(formula="y ~ x + s")
+    >>> rf.fit(df).transform(df).show()
```
Comment (MechCoder) [inline python/pyspark/ml/feature.py]: Not a PySpark comment, but is this expected behavior?

     >>> df = sqlContext.createDataFrame([
    ...     (1.0, 1.0, "a"),
    ...     (0.0, 2.0, "b"),
    ...     (0.0, 0.0, "c")
    ... ], ["y", "x", "s"])
    >>> rf.fit(df).transform(df).collect()
    [Row(y=1.0, x=1.0, s='a', features=DenseVector([1.0, 1.0, 0.0]), label=1.0),
     Row(y=0.0, x=2.0, s='b', features=DenseVector([2.0, 0.0, 1.0]), label=0.0),
     Row(y=0.0, x=0.0, s='c', features=SparseVector(3, {}), label=0.0)]

Comment (MechCoder) [inline python/pyspark/ml/feature.py]: Out of curiosity, in the test here (https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaSuite.scala#L104) why isn't the vector of dimension 4 (2 from the OneHotEncoder and two others), since there are three terms?

**Comment by MechCoder** [comment]
LGTM apart from those minor comments.

(I never knew a RFormula existed before, so thanks for that as well)

File: python/pyspark/ml/feature.py (hunk 2 above)
Comment (mengxr) [inline python/pyspark/ml/feature.py]: There are couple places to set parameters. Only checking here might not be sufficient. We need to address this issue separately.

File: python/pyspark/ml/feature.py (hunk 3 above)
Comment (mengxr) [inline python/pyspark/ml/feature.py]: Added one.

File: python/pyspark/ml/feature.py (hunk 4 above)
Comment (mengxr) [inline python/pyspark/ml/feature.py]: Yes, but I tried to create a small test case such that the behavior is well-defined, e.g., the ordering of string values. If we use `a`, `b`, and `c` with each appearing only once, how we determine the indices are not well defined. Should be a TODO item, but not part of this PR.

For the test, it is following DataFrames tests. It might be simpler to read as part of the documentation.

**Comment by mengxr** [comment]
test this please

**Comment by MechCoder** [comment]
thanks for the updates.

**Comment by mengxr** [comment]
Merged into master and branch-1.5.
